### PR TITLE
Fix resource leaking and make the code more testable

### PR DESCRIPTION
**Pull request by Vladislav-Kisliy**
`isConnected` method doesn't cover all cases. For example:
If a device has opened the ssh port however doesn't have netconf subsystem, a connection gets stuck for a long period of time.

### CODE REVIEW

**Comment by ydnath** [comment]
@peterjhill please review

File: src/main/java/net/juniper/netconf/Device.java [hunk 1]
```diff
@@ -61,48 +60,47 @@
     private static final int DEFAULT_NETCONF_PORT = 830;
     private static final int DEFAULT_TIMEOUT = 5000;
 
-    private String hostName;
-    private int port;
-    private int connectionTimeout;
-    private int commandTimeout;
+    private final JSch sshClient;
```
Comment (peterjhill) [inline src/main/java/net/juniper/netconf/Device.java:63]: good call making all these final

File: src/main/java/net/juniper/netconf/Device.java [hunk 2]
```diff
@@ -135,9 +133,9 @@ public Device(
         }
 
         this.netconfCapabilities = (netconfCapabilities != null) ? netconfCapabilities : getDefaultClientCapabilities();
-        this.helloRpc = createHelloRPC(this.netconfCapabilities);
+        helloRpc = createHelloRPC(this.netconfCapabilities);
 
-        this.sshClient = new JSch();
+        this.sshClient = (sshClient != null) ? sshClient : new JSch();
```
Comment (peterjhill) [inline src/main/java/net/juniper/netconf/Device.java:138]: love using ternary expressions..

File: src/main/java/net/juniper/netconf/Device.java [hunk 3]
```diff
@@ -189,8 +187,6 @@ private String createHelloRPC(List<String> capabilities) {
      */
     private NetconfSession createNetconfSession() throws NetconfException {
         if (!isConnected()) {
```
Comment (peterjhill) [inline src/main/java/net/juniper/netconf/Device.java:189]: it looks like we don't do anything interesting in this method anymore, now that you in line 138 do the 
`this.sshClient = (sshClient != null) ? sshClient : new JSch();`

should we make the createNeconfSession a no-op and deprecate it?

**Review by peterjhill — COMMENTED** [review]
this is great. I added some comments. The main one is whether or not we should make the createNetconfSession a no-op, since it no longer seems to do anything interesting.

File: src/main/java/net/juniper/netconf/Device.java (hunk 3 above)
Comment (Vladislav-Kisliy) [reply src/main/java/net/juniper/netconf/Device.java:189]: We use `createNetconfSession` in `connect()` method and sshClient initialization is a different thing than session creation. no-op means no-operations, correct?

Comment (peterjhill) [reply src/main/java/net/juniper/netconf/Device.java:189]: ah ha! when i was looking at the github ui, it looked like the create netconf session method code ended below  line 192.

got it!. 👍

**Review by peterjhill — APPROVED** [review]
Looks great!
 @ydnath looks good to me.. 

Thanks @Vladislav-Kisliy

File: src/main/java/net/juniper/netconf/Device.java (hunk 3 above)
Comment (Vladislav-Kisliy) [reply src/main/java/net/juniper/netconf/Device.java:189]: oh, yeah. GitHub sometimes is very confusing

**Pull request merged**